tests/test/closures.rs

```diff
@@ -0,0 +1,259 @@
+use super::*;
+
+#[test]
+fn closure_is_well_formed() {
+    test! {
+        program {
+            closure foo(self,) {}
+            closure bar(&self,) {}
+            closure baz(&mut self,) {}
+        }
+
+        goal {
+            WellFormed(foo)
+        } yields {
+            "Unique"
+        }
+        goal {
+            WellFormed(bar)
+        } yields {
+            "Unique"
+        }
+        goal {
+            WellFormed(baz)
+        } yields {
+            "Unique"
+        }
+    }
+}
+
+#[test]
+fn closure_is_sized() {
+    test! {
+        program {
+            #[lang(sized)]
+            trait Sized { }
+
+            closure foo(self,) {}
+        }
+
+        goal {
+            foo: Sized
+        } yields {
+            "Unique"
+        }
+    }
+}
+
+#[test]
+fn closure_is_copy() {
+    test! {
+        program {
+            #[lang(copy)]
+            trait Copy { }
+            impl<'a, T> Copy for &'a T {}
+            impl Copy for u8 {}
+            impl Copy for u16 {}
+            impl Copy for u32 {}
+
+            closure foo(self,) {}
+            closure bar(&self,) {}
+            closure baz(&mut self,) {}
+
+            closure foobuzz<'a>(self, a: u8, b: f32) -> u32 {
+                u8;
+                &'a u16;
+                &'a mut u32
+            }
+            closure foobar<'a>(self, a: u8, b: f32) -> u32 {
+                u8;
+                &'a u16
+            }
+        }
```

Comment:
can we add a test with generic types? maybe

```
closure with_ty<T>(self) { T; }
```

and then

```
forall<T> { with_ty<T>: Copy }
```

should not be true, but `forall<T> { if (T: Copy) { .. } }` should be provable.